# Tarea 1 Deep Learning

Francisco Rencoret (FranciscoRencoret1) - Raimundo Manterola (rmant)

In [44]:
#Definicion de librerias con la funciones que seran utilizadas por Keras.
import keras
from keras.layers import Activation, Dense, Flatten, Dropout
from keras.models import Sequential
from keras.layers.convolutional import Convolution2D, ZeroPadding2D
from keras.layers.normalization import BatchNormalization
from keras.layers.pooling import MaxPooling2D


In [45]:
#Definicion de contenedor y primera capa de AlexNet.
modelAlexNet = Sequential()
modelAlexNet.add(ZeroPadding2D((2,2), input_shape=(224, 224, 3)))
modelAlexNet.add(Convolution2D(96, (11,11), strides=(4,4), padding="valid"))
modelAlexNet.add(Activation(activation="relu"))
output_shape_first_layer = modelAlexNet.output_shape
print('Los mapas de activación de la capa 1 de convolción tienen tamaño: {}'.format(output_shape_first_layer))
modelAlexNet.add(BatchNormalization())
modelAlexNet.add(MaxPooling2D((3,3), strides=(2,2)))

Los mapas de activación de la primera capa de convolción tienen tamaño: (None, 55, 55, 96)


In [46]:
#Definicion de la segunda capa de AlexNet.
modelAlexNet.add(ZeroPadding2D((2,2)))
modelAlexNet.add(Convolution2D(256, (5, 5), padding="valid"))
modelAlexNet.add(Activation(activation="relu"))
output_shape_second_layer = modelAlexNet.output_shape
print('Los mapas de activación de la capa 2 de convolción tienen tamaño: {}'.format(output_shape_second_layer))
modelAlexNet.add(BatchNormalization())
modelAlexNet.add(MaxPooling2D((3,3), strides=(2,2)))

Los mapas de activación de la segunda capa de convolción tienen tamaño: (None, 27, 27, 256)


Actividad 1

Notamos que la primera capa de AlexNet es de tamaño (55, 55, 96). Como podemos ver, nuestra primera capa tiene las mismas dimensiones:

In [47]:
output_shape_first_layer

(None, 55, 55, 96)

La segunda capa de AlexNet es de tamanño (27, 27, 256). Nuestra segunda capa tiene las mismas dimensiones

In [48]:
output_shape_second_layer

(None, 27, 27, 256)

Ahora, vamos a mostrar un resumen de la red creada

In [49]:
modelAlexNet.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
zero_padding2d_19 (ZeroPaddi (None, 228, 228, 3)       0         
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 55, 55, 96)        34944     
_________________________________________________________________
activation_19 (Activation)   (None, 55, 55, 96)        0         
_________________________________________________________________
batch_normalization_9 (Batch (None, 55, 55, 96)        384       
_________________________________________________________________
max_pooling2d_12 (MaxPooling (None, 27, 27, 96)        0         
_________________________________________________________________
zero_padding2d_20 (ZeroPaddi (None, 31, 31, 96)        0         
_________________________________________________________________
conv2d_20 (Conv2D)           (None, 27, 27, 256)       614656    
__________

Podemos notar que en la primera capa, teniamos filtros de tamaño 11 x 11 x 3 + 1 pesos (11 x 11 con profundidad 3 + el bias). Ahora, como la primera capa tiene 96 mapas de activacion, vamos a tener 96 de estos filtros, por lo que en la primera capa se tiene (11 x 11 x 3 + 1) x 96 = 34,944 pesos. Podemos ver que esto coincide con el valor de la segunda fila de la tabla.

La segunda capa tiene filtros de 5 x 5 x 96 + 1 (5 x 5 con profundidad 96 + un bias) y tenemos 256 mapas de activacion, por lo que en total la capa tiene (5 x 5 x 96 + 1) x 256 = 614,656. Esto coincide con el valor de la septima fila de la tabla.

Por otra parte, podemos notar que el batch normaliation en ambos casos calcula 4 parametros. Como la normalización consiste en:

$$\begin{equation}
\hat{a} =  \frac{x - \mu}{\sigma}
\end{equation}$$

y luego

$$\begin{equation}
\hat{x} =  \varphi \hat{a} + \beta
\end{equation}$$

la red debe calcular y almacenar $\mu$ y $\sigma$ (parametros no entrenables). Por otro lado, debe calcular y entrenar a $\varphi$ y $\beta$, por lo que por cada mapa de activacion almacena 4 pesos pero solo entrena 2. En total, primer batch tiene 96 x 4 = 384 pesos (192 entrenables) y el segundo 256 x 4 = 1024 (512 entrenables).

En resumen:
Pesos: 34,944 + 384 + 614,656 + 1024 = 651,008
Pesos entrenables: 34,944 + 192 + 614,656 + 512 = 650,304

Actividad 2

In [50]:
#Definicion de la tercera capa de AlexNet
modelAlexNet.add(ZeroPadding2D((1,1)))
modelAlexNet.add(Convolution2D(384, (3, 3), strides=(1,1), padding="valid"))
modelAlexNet.add(Activation(activation="relu"))
output_shape_third_layer = modelAlexNet.output_shape
print('Los mapas de activación de la capa 3 de convolción tienen tamaño: {}'.format(output_shape_third_layer))

Los mapas de activación de la tecera capa de convolción tienen tamaño: (None, 13, 13, 384)


Verificamos que tenga las mismas dimensiones

In [51]:
output_shape_third_layer

(None, 13, 13, 384)

Volvemos a revisar el resumen de la red

In [52]:
modelAlexNet.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
zero_padding2d_19 (ZeroPaddi (None, 228, 228, 3)       0         
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 55, 55, 96)        34944     
_________________________________________________________________
activation_19 (Activation)   (None, 55, 55, 96)        0         
_________________________________________________________________
batch_normalization_9 (Batch (None, 55, 55, 96)        384       
_________________________________________________________________
max_pooling2d_12 (MaxPooling (None, 27, 27, 96)        0         
_________________________________________________________________
zero_padding2d_20 (ZeroPaddi (None, 31, 31, 96)        0         
_________________________________________________________________
conv2d_20 (Conv2D)           (None, 27, 27, 256)       614656    
__________

Actividad 3

In [53]:
#Definicion de la cuarta capa de AlexNet
modelAlexNet.add(ZeroPadding2D((1,1)))
modelAlexNet.add(Convolution2D(384, (3, 3), strides=(1,1), padding="valid"))
modelAlexNet.add(Activation(activation="relu"))
output_shape_fourth_layer = modelAlexNet.output_shape
print('Los mapas de activación de la capa 4 de convolción tienen tamaño: {}'.format(output_shape_fourth_layer))

Los mapas de activación de la cuarta capa de convolción tienen tamaño: (None, 13, 13, 384)


In [54]:
#Definicion de la quinta capa de AlexNet
modelAlexNet.add(ZeroPadding2D((1,1)))
modelAlexNet.add(Convolution2D(256, (3, 3), strides=(1,1), padding="valid"))
modelAlexNet.add(Activation(activation="relu"))
output_shape_fifth_layer = modelAlexNet.output_shape
print('Los mapas de activación de la capa 5 de convolción tienen tamaño: {}'.format(output_shape_fifth_layer))
modelAlexNet.add(MaxPooling2D((3,3), strides=(2,2)))

Los mapas de activación de la quinta capa de convolción tienen tamaño: (None, 13, 13, 256)


Nuevamente verificamos que las dimensiones son correctas. Al igual que la capa 3, las dimensiones de la capa 4 son (13, 13, 384). La quinta capa tiene dimensiones (13, 13, 256). Viendo las dimensiones de AlexNet, notamos que coinciden. Ahora, revisemos el resumen del modelo para ver la cantidad de parametros:

In [55]:
modelAlexNet.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
zero_padding2d_19 (ZeroPaddi (None, 228, 228, 3)       0         
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 55, 55, 96)        34944     
_________________________________________________________________
activation_19 (Activation)   (None, 55, 55, 96)        0         
_________________________________________________________________
batch_normalization_9 (Batch (None, 55, 55, 96)        384       
_________________________________________________________________
max_pooling2d_12 (MaxPooling (None, 27, 27, 96)        0         
_________________________________________________________________
zero_padding2d_20 (ZeroPaddi (None, 31, 31, 96)        0         
_________________________________________________________________
conv2d_20 (Conv2D)           (None, 27, 27, 256)       614656    
__________

Podemos notar que la cuarta capa tiene más parámetros que la tercera capa, porque a pesar de tener la misma estrucutra (filtros de igual tamaño pero distinta profundidad, mapas de activación de igual tamaño), vemos que su input es distinto. La tercera recibe un input de (13, 13, 256), mientras que la cuarta recibe un input de (13, 13, 384). Por lo tanto, los filtros van a tener profundidades distintas lo que se traduce en una distinta cantidad de parametros.

Por otra parte, notamos que la quinta capa recibe el mismo input que la cuarta, pero como la quinta tiene menos mapas de activación, la cantidad de parametros que debe manejar es menor. 

Por último, notamos que como no incluimos normalización de los batches, la cantidad de parametros no entrenables no cambia.

Actividad 4

Actualmente tenemos la matriz como un tensor de 3D. Como ahora tenemos que agregar las capas densas para clasificar, debemos tener la matríz como un vector de una dimensión, para luego poder hacer correctamente las conexiones con la capa densa. Para eso, usamos el comando Flatten

In [56]:
modelAlexNet.add(Flatten())

In [61]:
modelAlexNet.output_shape

(None, 9216)

Notamos que esto efectivamente hizo lo que necesitabamos, porque ahora el modelo tiene dimensiones (1, 9216), justo lo que necesitabamos 